File: AppMovies/app/src/main/java/com/svmdev/appmovies/view/adapter/MovieAdapter.java
package com.svmdev.appmovies.view.adapter;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.svmdev.appmovies.R;
import com.svmdev.appmovies.data.model.Movie;

import java.util.ArrayList;

public class MovieAdapter extends ArrayAdapter<Movie> {

    Context context;

    public MovieAdapter(Context context, ArrayList<Movie> movies) {
        super(context, 0, movies);
        this.context = context;
    }

    public View getView(int pos, View view, ViewGroup parent) {

        //make this cell is recyclable
        if (view == null) {
            view = LayoutInflater.from(getContext()).inflate(R.layout.adapter_movie, parent, false);
        }

        Movie movie = this.getItem(pos);

        ImageView picture = view.findViewById(R.id.adapt_movie_pic);
        TextView title = view.findViewById(R.id.adapt_movie_title);
        TextView date = view.findViewById(R.id.adapt_movie_release);

        title.setText(movie.getTitle());
        date.setText(movie.getFormatedReleaseDate());

        if(!movie.getPosterPath().isEmpty()) {
            Picasso.with(context).load(movie.getPosterPath())
                    .placeholder(R.drawable.download)
                    .error(R.drawable.no_image)
                    .into(picture, new com.squareup.picasso.Callback() {

                        @Override
                        public void onSuccess() {
                        }

                        @Override
                        public void onError() {
                        }


                    });
        } else {
            picture.setImageResource(R.drawable.no_image);
        }

        return view;
    }

}

File: app/src/main/java/com/brunodiegom/movies/view/MovieAdapter.kt
package com.brunodiegom.movies.view

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.brunodiegom.movies.R
import com.brunodiegom.movies.extension.toDate
import com.brunodiegom.movies.model.Movie
import com.brunodiegom.movies.model.Movie.Companion.DIFF_CALLBACK
import com.brunodiegom.movies.model.Movie.Companion.THUMBNAIL_BASE_URL
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.adapter_movie.view.poster_thumbnail
import kotlinx.android.synthetic.main.adapter_movie.view.release_date
import kotlinx.android.synthetic.main.adapter_movie.view.title

/**
 * [RecyclerView.Adapter] to apply the [Movie] content on the list.
 */
class MovieAdapter(
    private val listener: MovieAdapterItemListener
) : PagedListAdapter<Movie, MovieAdapter.MovieAdapterViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        MovieAdapterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.adapter_movie, parent, false))

    override fun onBindViewHolder(holder: MovieAdapterViewHolder, position: Int) {
        getItem(position)?.let { movie ->
            holder.view.apply {
                setOnClickListener { listener.onClick(movie) }
                title.text = movie.title
                release_date.text = movie.releaseDate.toDate()
                Picasso.get()
                    .load("$THUMBNAIL_BASE_URL${movie.posterUrl}")
                    .fit()
                    .placeholder(R.mipmap.poster_placeholder)
                    .error(R.mipmap.poster_placeholder)
                    .into(poster_thumbnail)
            }
        }
    }

    /**
     * [RecyclerView.ViewHolder] to contain the adapter view.
     */
    class MovieAdapterViewHolder(val view: View) : RecyclerView.ViewHolder(view)

    /**
     * Item selection interface.
     */
    interface MovieAdapterItemListener {
        /**
         * How implements this method will know when an [Movie] was selected, and which one,
         */
        fun onClick(movie: Movie)
    }
}
